proposals/0540-vision-for-layout-conformance-parity.md

```diff
@@ -0,0 +1,238 @@
+
+## Synchronous Layout
+
+While not a W3C conformance issue, React Native doesn’t expose the DOMs methods of synchronous layout, which bleeds into differences in the React programming model. React has both “effects“ and ”layout effects”. On the web, effects (e.g. code in `useEffect`) are executed after the component has been laid out and painted. Layout effects are executed directly after DOM mutation, and block painting. In React Native, both effects and layout effects are executed in parallel with layout and painting. This limits the ability for components to synchronously measure components, or programmatically respond to layout changes, without relying on first committing them to screen. Layout computation is not thread affined today (it can happen off the UI thread). A solution for this is to allow synchronously running layout on the JS thread. This was previously attempted, but had (possibly mitigable) performance concerns when measured.
+
+## Choosing the capabilities to build
+
+React Native does not implement the full set of browser CSS features. Apart from higher level concepts like stylesheets, selectors, and cascading (which can largely be supplanted by CSS-in-JS solutions shared between RN and browser), React Native doesn’t implement all style properties.
+
+It’s clear there is a desire for more capabilities, but there is a question of which should be added. React Native should enable skills to be transferable, offer a stellar developer experience, and enable substantial code-sharing (while being leaner than a browser). A natural conflict is whether to prioritize building for compatibility (existing usages), or for developer experience. [Adoption of conformant layout](#restriction-to-conformant-trees) is dependent on building out features developers would like to use, in their existing React Native projects. 
+
+With the initial focus of building for developer experience, instead of compatibility, we need input on what the community cares about. We can derive a lot of this from asking directly, but should be opportunistic in our choices, accounting for implementation complexity.
+
+## Mitigating bloat to heap usage
+
+Yoga internally represents a stylesheet as a packed structure of all possible properties. This structure is already 192 bytes, scaling with each additional Yoga Node. As we add more styles, this could expand. Doubling the number of styles on a tree of 20,000 nodes would increase memory consumption by almost 4MB, a huge amount for some of the devices React Native targets. We should measure, and potentially switch YGStyle to a data structure which is pay-for-play in number of properties. This shift would be performance sensitive, but theoretically quick data structures exist (e.g. a bitfield header with a small-vector for larger values). 
```

Comment:
I wonder if we got RN to handle the cascading of margin, padding etc. properties, if we could reduce this by a fair bit

RN already stores each value (each direction, horizontal, vertical, all) in an array, then applies each to the yoga node - https://github.com/facebook/react-native/blob/a9bed8e75d9b9613c5fcb69436a2d3af763f456d/React/Views/RCTShadowView.m#L99

We could literally just do the resolution here and only set the directional ones on the yoga node

Reply:
I did something a bit similar for aliased props in Fabric recently https://github.com/facebook/react-native/pull/35342. But I discovered the Fabric design for ShadowNode property storage runs into the same issue as YGStyle, and there are even two separate copies of YGStyle (one in ShadowNode props and the other on the Yoga node) so we are IIRC at 1.2KB of empty props per-ShadowNode already.

I think fixing that up in Fabric is a goal some folks are taking up next half, and it could deliver even bigger wins than just YGStyle itself, though YGStyle currently takes up a big portion of per-shadownode memory usage.

Reply:
If the cascade was completely handled on the RN side, so we could also just use direct access on the properties without having these helpers to compute margins etc. - and might get a slight perf boost too